src/libmongoc/tests/unified/entity-map.c

```diff
@@ -0,0 +1,603 @@
+   }
+
+   database_entity = entity_map_get (entity_map, database_id, error);
+   database = (mongoc_database_t *) database_entity->value;
+   entity->value = mongoc_database_get_collection (database, collection_name);
+   ret = true;
+done:
+   bson_free (collection_name);
+   bson_free (database_id);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+mongoc_session_opt_t *
+session_opts_new (bson_t *bson, bson_error_t *error)
+{
+   bool ret = false;
+   mongoc_session_opt_t *opts = NULL;
+   bson_parser_t *parser;
+   bool *causal_consistency = NULL;
+
+   parser = bson_parser_new ();
+   bson_parser_bool_optional (parser, "causalConsistency", &causal_consistency);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   opts = mongoc_session_opts_new ();
+   if (causal_consistency) {
+      mongoc_session_opts_set_causal_consistency (opts, *causal_consistency);
+   }
+
+done:
+   bson_parser_destroy (parser);
+   bson_free (causal_consistency);
+   if (!ret) {
+      mongoc_session_opts_destroy (opts);
+      return NULL;
+   }
+   return opts;
+}
+
+entity_t *
+entity_session_new (entity_map_t *entity_map, bson_t *bson, bson_error_t *error)
+{
+   bson_parser_t *parser;
+   entity_t *entity;
+   entity_t *client_entity;
+   mongoc_client_t *client;
+   char *client_id = NULL;
+   bson_t *session_opts_bson = NULL;
+   mongoc_session_opt_t *session_opts = NULL;
+   bool ret = false;
+
+   entity = entity_new ("session");
+   parser = bson_parser_new ();
+   bson_parser_utf8 (parser, "id", &entity->id);
+   bson_parser_utf8 (parser, "client", &client_id);
+   bson_parser_doc (parser, "sessionOptions", &session_opts_bson);
+   if (!bson_parser_parse (parser, bson, error)) {
+      goto done;
+   }
+
+   client_entity = entity_map_get (entity_map, client_id, error);
+   client = (mongoc_client_t *) client_entity->value;
+   session_opts = session_opts_new (session_opts_bson, error);
+   if (!session_opts) {
+      goto done;
+   }
+   entity->value = mongoc_client_start_session (client, session_opts, error);
+   if (!entity->value) {
+      goto done;
+   }
+   ret = true;
+done:
+   mongoc_session_opts_destroy (session_opts);
+   bson_free (client_id);
+   bson_destroy (session_opts_bson);
+   bson_parser_destroy (parser);
+   if (!ret) {
+      entity_destroy (entity);
+      return NULL;
+   }
+   return entity;
+}
+
+/* Caveat: The spec encourages, but does not require, that entities are defined
+ * in dependency order:
+ * "Test files SHOULD define entities in dependency order, such that all
+ * referenced entities (e.g. client) are defined before any of their dependent
+ * entities (e.g. database, session)."
+ * If a test ever does break this pattern (flipping dependency order), that can
+ * be solved by creating C objects lazily in entity_map_get.
+ * The current implementation here does the simple thing and creates the C
```

Comment:
Seems reasonable!  If a test ever defined entities out of order, could we get around that by, like, sorting the dependencies before parsing them?  Parsing all clients first, then all databases, etc etc.  For a simple but inefficient solution we could loop over the list of entities four separate times, selectively parsing one entity type on each loop through.

Reply:
Doing multiple passes seems like a straightforward solution. I don't imagine there will ever be more then 100 entities in a test file. So four passes seems totally fine.